common/src/sphere_collider.rs

```diff
@@ -0,0 +1,800 @@
+    // Note that discriminant can be negative, which allows this function to return NaN when there is no solution.
+    Some(constant_term / (-half_linear_term + discriminant.sqrt()))
+}
+
+/// Converts from t-u-v coordinates to x-y-z coordinates. t-u-v coordinates are a permuted version of x-y-z coordinates.
+/// `t_axis` determines which of the three x-y-z coordinates corresponds to the t-coordinate. This function works with
+/// any indexable entity with at least three entries. Any entry after the third entry is ignored.
+fn tuv_to_xyz<T: std::ops::IndexMut<usize, Output = N>, N: Copy>(t_axis: usize, tuv: T) -> T {
+    let mut result = tuv;
+    (
+        result[t_axis],
+        result[(t_axis + 1) % 3],
+        result[(t_axis + 2) % 3],
+    ) = (result[0], result[1], result[2]);
+    result
+}
+
+/// Converts a single coordinate from dual coordinates in the Klein-Beltrami model to an integer coordinate
+/// suitable for voxel lookup. Margins are included. Returns `None` if the coordinate is outside the chunk.
+#[inline]
+fn dual_to_voxel(ctx: &ChunkShapeCastingContext, dual_coord: f32) -> Option<usize> {
+    let floor_grid_coord = (dual_coord * ctx.dual_to_grid_factor).floor();
+
+    if !(floor_grid_coord >= 0.0 && floor_grid_coord < ctx.dimension as f32) {
+        None
+    } else {
+        Some(floor_grid_coord as usize + 1)
+    }
+}
+
+/// Converts a single coordinate from grid coordinates to dual coordiantes in the Klein-Beltrami model. This
+/// can be used to find the positions of voxel gridlines.
+#[inline]
+fn grid_to_dual(ctx: &ChunkShapeCastingContext, grid_coord: usize) -> f32 {
+    grid_coord as f32 / ctx.dual_to_grid_factor
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        collision::VoxelAABB,
+        dodeca::Vertex,
+        graph::NodeId,
+        node::{ChunkId, VoxelData},
+    };
+
+    use super::*;
+    use approx::*;
+
+    /// Helper structure used along with a ray to generate a `ChunkShapeCastingContext`
+    struct TestShapeCastContext {
+        collider_radius: f32,
+        dimension: usize,
+        dual_to_grid_factor: f32,
+        voxel_data: VoxelData,
+        transform: na::Matrix4<f32>,
+    }
+
+    impl TestShapeCastContext {
+        fn new(collider_radius: f32) -> Self {
+            // Use an arbitrary transformation
+            let transform = na::Rotation3::from_euler_angles(0.1, 0.2, 0.3).to_homogeneous()
+                * math::translate_along(&na::Vector3::new(0.5, 0.3, 0.2));
+            let dimension: usize = 12;
+
+            let mut test_ctx = TestShapeCastContext {
+                collider_radius,
+                dimension,
+                dual_to_grid_factor: Vertex::dual_to_chunk_factor() as f32 * dimension as f32,
+                voxel_data: VoxelData::Solid(Material::Void),
+                transform,
+            };
+
+            // Populate voxels. Consists of a single cube with grid coordinates from (1, 1, 1) to (2, 2, 2)
```

Comment:
What space are the coordinates in this comment? The context suggests voxel coordinates, but only a single voxel is set, not a box of eight.

Reply:
They're in "grid coordinates". the voxel coordinates would just be (2, 2, 2).

Reply:
Is that scheme used anywhere else?

Reply:
It was one of the coordinate systems discussed in Discord, and it's used for the iterators in `VoxelAABB`.

Reply:
Let's revisit this in the future I guess. Probably mostly a doc problem.